### PR TITLE
Feature/avatar upload

### PR DESCRIPTION
**Pull request by Schnoop**
https://github.com/FrankProjects/UltimateWarfare/issues/8

### CODE REVIEW

**Comment by frank9999** [comment]
What is your opinion about storing the avatar as base64encoded string in the database? I'm using multi-server deployments, so persisting user uploads in 1 directory on 1 server will cause issues. Several deployment tools also create a new checkout in a separate directory, which will cause the avatars to break. I can work around these issues by making the avatar upload directory a shared NFS share or use a shared object store... Storing the image in the database seems like the easiest solution, but it may cause performance issues and increase database disk usage/load. @Schnoop

**Comment by Schnoop** [comment]
Yeah - that would be a good option. If the database should get in trouble at some point in the future it should not be a problem to move the images back to files and think about other deployment options. Maybe a S3 storage is another option to store the user generated images.

**Comment by Schnoop** [comment]
Changed implementation to store image in database.

File: composer.json [hunk 1]
```diff
@@ -6,8 +6,9 @@
     "require": {
         "php": "^8.1",
         "ext-gd": "*",
-        "ext-json": "*",
         "ext-intl": "*",
+        "ext-json": "*",
+        "ext-imagick": "*",
```
Comment (frank9999) [inline composer.json:11]: Any specific reason you prefer imagick over gd? I'm not opposed using imagick but it adds more dependencies to the project and other peoples hosting environment.

```
sudo apt-get install php8.1-imagick
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  fonts-droid-fallback fonts-noto-mono fonts-urw-base35 ghostscript gsfonts imagemagick-6-common libaom3
  libavahi-client3 libavahi-common-data libavahi-common3 libcups2 libdav1d5 libde265-0 libfftw3-double3 libgomp1
  libgs9 libgs9-common libheif1 libidn12 libijs-0.35 libjbig2dec0 liblcms2-2 liblqr-1-0 libltdl7 libmagickcore-6.q16-6
  libmagickwand-6.q16-6 libopenjp2-7 libpaper-utils libpaper1 libwebpdemux2 libwebpmux3 libx265-199 poppler-data
Suggested packages:
  fonts-noto fonts-freefont-otf | fonts-freefont-ttf fonts-texgyre ghostscript-x cups-common libfftw3-bin libfftw3-dev
  liblcms2-utils libmagickcore-6.q16-6-extra poppler-utils fonts-japanese-mincho | fonts-ipafont-mincho
  fonts-japanese-gothic | fonts-ipafont-gothic fonts-arphic-ukai fonts-arphic-uming fonts-nanum
Recommended packages:
  ttf-dejavu-core
The following NEW packages will be installed:
  fonts-droid-fallback fonts-noto-mono fonts-urw-base35 ghostscript gsfonts imagemagick-6-common libaom3
  libavahi-client3 libavahi-common-data libavahi-common3 libcups2 libdav1d5 libde265-0 libfftw3-double3 libgomp1
  libgs9 libgs9-common libheif1 libidn12 libijs-0.35 libjbig2dec0 liblcms2-2 liblqr-1-0 libltdl7 libmagickcore-6.q16-6
  libmagickwand-6.q16-6 libopenjp2-7 libpaper-utils libpaper1 libwebpdemux2 libwebpmux3 libx265-199 php8.1-imagick
  poppler-data
0 upgraded, 34 newly installed, 0 to remove and 0 not upgraded.
Need to get 294 kB/27.6 MB of archives.
After this operation, 106 MB of additional disk space will be used.
```

File: .gitignore [hunk 2]
```diff
@@ -11,6 +11,7 @@
 .idea
 /config/secrets/dev/
 .ddev/
+/public/uploads/avatars/
```
Comment (frank9999) [inline .gitignore:14]: Is this still needed as images are stored in the database now?

File: src/Entity/User.php [hunk 3]
```diff
@@ -176,6 +177,25 @@ public function setEmail(string $email): void
...
+    public function setAvatar(mixed $avatar): void
+    {
+        $this->avatar = $avatar;
+    }
+
+    public function hasAvatar(): bool
+    {
+        $stats = fstat($this->avatar);
+        if ($stats !== false) {
+            return $stats[7] > 0;
```
Comment (frank9999) [inline src/Entity/User.php:194]: Is $stats['size'] also working? It allows developers to quicker read and understand what this function is doing.

File: src/Form/UploadAvatarType.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        $builder
+            ->setAction('/game/account/upload_avatar')
+            ->add(
+                'avatar',
+                FileType::class,
+                [
+                    'mapped' => false,
+                    'label' => 'label.avatar',
+                    'constraints' => [
+                        new Image([
```
Comment (frank9999) [inline src/Form/UploadAvatarType.php:28]: You probably need to add a use statement for this Image object? I get a ClassNotFoundException.

> Attempted to load class "Image" from namespace "Symfony\Component\Validator\Constraints"

File: templates/game/profile.html.twig [hunk 5]
```diff
@@ -8,8 +8,17 @@
...
         <tr>
-            <td width="150">Name:</td>
-            <td width="350">{{ profilePlayer.name }}</td>
+            <td width="150" valign="middle">Name:</td>
+            <td width="350">
+                <table class="table text-center">
+                    <tr>
+                        <td style="border:0" valign="middle">{{ profilePlayer.name }}</td>
+                        {% if profilePlayer.getUser.avatar != '' %}
+                            <td style="border:0" align="right"><img src="{{ asset('uploads/avatars/' ~ profilePlayer.getUser.avatar) }}"/></td>
```
Comment (frank9999) [inline templates/game/profile.html.twig:17]: This still needs rewriting to the base64encoded image like in editAccount.html.twig?

**Review by frank9999 — COMMENTED** [review]
Thanks for this merge request! I left some review notes :)

File: src/Form/UploadAvatarType.php (hunk 4 above)
Comment (Schnoop) [reply src/Form/UploadAvatarType.php:28]: Maybe you missed that i've added the validator package in the composer.json file. A composer install might fix that problem for you.

Comment (frank9999) [reply src/Form/UploadAvatarType.php:28]: Ah missed that one! Forgot to composer install after installing the dependencies :)

File: composer.json (hunk 1 above)
Comment (Schnoop) [reply composer.json:11]: I failed to find a way with the GD library. I think it is a good idea to stick to Imagick which works well - even though it's a dependency

**Pull request merged**